#!/usr/bin/env Rscript 

# import shell arguments
args = commandArgs(trailingOnly=TRUE)

# test if there is at least one argument: if not, return an error
if (length(args)!=3) {
  stop("\nUsage: Rscript nanograd3.R /path/to/annotation.gtf /path/to/coverage.txt /path/to/output.txt", call.=FALSE)
}

library(GenomicFeatures)
library(rtracklayer)
library(tidyverse)

# interactive args for testing (Melbourne iMac)
# args <- c("/Users/asethi/localGadiData/2022-03-15_test_nanograd3/Mus_musculus.GRCm39.104.chr.gtf", "/Users/asethi/localGadiData/2022-03-15_test_nanograd3/coverage.txt", "/Users/asethi/localGadiData/2022-03-15_test_nanograd3/nano_out.txt")

################################################################################
################################################################################
################################################################################

# start

tstart <- print(paste("start time is", as.character(Sys.time())))

##################################################
##################################################
##################################################

# process the annotation and make a reference table with transcript structure information

# read in reference transcripts
gtf <- makeTxDbFromGFF(file=args[1], format = "gtf")

# make an exon database from the reference transcripts
exons <- exonsBy(gtf, "tx", use.names=TRUE)

# convert the transcripts to a tibble
exons_tib <- as_tibble(as(exons, "data.frame"))

# fetch the length of each transcript segment from the gtf
txlen <- transcriptLengths(gtf, with.cds_len=TRUE, with.utr5_len=TRUE, with.utr3_len=TRUE) %>%
  as_tibble() %>%
  mutate(diff = tx_len - cds_len - utr5_len - utr3_len) %>%
  dplyr::rename(transcript_id = tx_name) %>%
  dplyr::select(-tx_id, -nexon, -gene_id, -diff)

# the last command doesn't store biotype, so we read in the gtf again using another package
tx_biotype <- rtracklayer::import(args[1]) %>%
  as_tibble() %>%
  dplyr::select(transcript_id, transcript_biotype, gene_name, gene_id) %>%
  na.omit() %>%
  dplyr::distinct()

# merge the biotypes with the transcript segment lengths
merged_metadata <- inner_join(tx_biotype, txlen, by = "transcript_id")

##################################################
##################################################
##################################################

# process the bam coverage track generated by the shell script 
coverage <- read_tsv(args[2], col_names = F, col_types = "fii") %>% 
  rename(tx = 1, crd = 2, cov = 3) %>% 
  group_by(tx) %>% 
  mutate(tx_cov = max(cov))

##########

# as a test, plot the coverage distribution per transcript 

cov_dist <- coverage %>% select(tx, tx_cov) %>% unique()
##########



